source/utils/algorithms.hpp

```diff
@@ -13,6 +13,6 @@ using RoutingTable = std::unordered_map<Id, MapWeakPtr<ILink, int>>;
 // Unlike standard BFS that processes nodes one by one, this processes all nodes
 // at the current distance level together. So each iteration is a processing of
 // all devices at a certain distance (wavefront)
-RoutingTable bfs(std::shared_ptr<IRoutingDevice>& start_device);
+RoutingTable bfs(std::shared_ptr<IDevice>& start_device);
```

Comment:
It looks like `IRoutingDevice` is enought abstraction for `bfs`:

```suggestion
RoutingTable bfs(std::shared_ptr<IRoutingDevice>& start_device);
```